index: Optimize generatePackList

greatroar opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Speeds up generating JSON from indexes, and especially cuts down the allocation rate. By using a []int instead of a map from ID's to pointers, it's also easier on the GC.

```
name                 old time/op    new time/op    delta
EncodeIndex/100-8      1.56ms ± 2%    1.40ms ± 3%  -10.38%  (p=0.000 n=10+9)
EncodeIndex/1000-8     14.5ms ± 2%    12.5ms ± 4%  -13.56%  (p=0.000 n=9+10)
EncodeIndex/10000-8     120ms ± 2%     110ms ± 5%   -8.38%  (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
EncodeIndex/100-8       306kB ± 1%     260kB ± 1%  -15.01%  (p=0.000 n=10+10)
EncodeIndex/1000-8     3.69MB ±11%    2.91MB ± 4%  -21.12%  (p=0.000 n=10+10)
EncodeIndex/10000-8    35.9MB ±11%    31.0MB ±23%  -13.60%  (p=0.015 n=10+10)

name                 old allocs/op  new allocs/op  delta
EncodeIndex/100-8       3.39k ± 0%     2.37k ± 0%  -30.08%  (p=0.000 n=10+10)
EncodeIndex/1000-8      32.6k ± 0%     22.9k ± 0%  -29.81%  (p=0.000 n=10+10)
EncodeIndex/10000-8      326k ± 0%      229k ± 0%  -29.80%  (p=0.000 n=10+10)
```

The bulk of the allocation rate improvement comes from just removing the debug.Log calls: every one of those copied a restic.ID to the heap.

The remaining allocations are mostly within the JSON encoder.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

#838 discusses JSON performance. (Any other format would still have to do this transformation, though.)

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

greatroar commented:
I also tried changing debug.Log to

```
func Log(f string, args ...interface{}) {
	if !opts.isEnabled {
		return
	}
	doLog(f, args...)
}

func doLog(f string, args ...interface{}) {
	/* ... */
}
```

While debug.Log can then be inlined, the allocation still happens.

MichaelEischer commented on internal/index/index.go:
```diff
@@ -317,30 +317,23 @@ type blobJSON struct {
 // generatePackList returns a list of packs.
-func (idx *Index) generatePackList() ([]*packJSON, error) {
-	list := []*packJSON{}
-	packs := make(map[restic.ID]*packJSON)
```
I'm somewhat unsure whether we should get rid of the map. Strictly speaking the index implementation does not guarantee a 1:1 mapping between pack IDs and the packIndex. If a pack is contained in multiple indexes, then after merging multiple indexes, the same pack ID might correspond to multiple packIndexes. However, `MasterIndex.Save`, which is the main user of `generatePackList`, that won't be a problem.

MichaelEischer replied:
What would be the benefit of just removing the debug.Log calls?

greatroar replied:
```
name                 old time/op    new time/op    delta
EncodeIndex/100-8      1.56ms ± 2%    1.50ms ± 1%   -3.91%  (p=0.000 n=10+9)
EncodeIndex/1000-8     14.5ms ± 2%    13.3ms ± 2%   -8.38%  (p=0.000 n=9+9)
EncodeIndex/10000-8     120ms ± 2%     116ms ± 2%   -3.66%  (p=0.000 n=10+9)

name                 old alloc/op   new alloc/op   delta
EncodeIndex/100-8       306kB ± 1%     276kB ± 1%   -9.91%  (p=0.000 n=10+10)
EncodeIndex/1000-8     3.69MB ±11%    2.98MB ± 5%  -19.19%  (p=0.000 n=10+10)
EncodeIndex/10000-8    35.9MB ±11%    30.3MB ±17%  -15.58%  (p=0.000 n=10+10)

name                 old allocs/op  new allocs/op  delta
EncodeIndex/100-8       3.39k ± 0%     2.48k ± 0%  -27.05%  (p=0.000 n=10+10)
EncodeIndex/1000-8      32.6k ± 0%     23.9k ± 0%  -26.73%  (p=0.000 n=10+10)
EncodeIndex/10000-8      326k ± 0%      239k ± 0%  -26.73%  (p=0.000 n=10+10)
```

But there's an intermediate option, where the list still changes from []*packJSON to []packJSON. Let me try that.

MichaelEischer commented:
> While debug.Log can then be inlined, the allocation still happens.

I guess the escape analysis isn't able to fully analyze the vararg. If I remember correctly, then the allocation should even happen, if `Log` is empty.

greatroar committed:
index: Optimize generatePackList

name                 old time/op    new time/op    delta
EncodeIndex/100-8      1.56ms ± 2%    1.48ms ± 3%   -5.37%  (p=0.000 n=10+10)
EncodeIndex/1000-8     14.5ms ± 2%    13.1ms ± 2%   -9.49%  (p=0.000 n=9+10)
EncodeIndex/10000-8     120ms ± 2%     116ms ± 2%   -3.58%  (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
EncodeIndex/100-8       306kB ± 1%     275kB ± 1%  -10.28%  (p=0.000 n=10+10)
EncodeIndex/1000-8     3.69MB ±11%    2.88MB ± 5%  -22.07%  (p=0.000 n=10+9)
EncodeIndex/10000-8    35.9MB ±11%    31.9MB ±10%  -11.13%  (p=0.005 n=10+10)

name                 old allocs/op  new allocs/op  delta
EncodeIndex/100-8       3.39k ± 0%     2.39k ± 0%  -29.61%  (p=0.000 n=10+10)
EncodeIndex/1000-8      32.6k ± 0%     22.9k ± 0%  -29.63%  (p=0.000 n=10+9)
EncodeIndex/10000-8      326k ± 0%      229k ± 0%  -29.71%  (p=0.000 n=10+10)

The bulk of the allocation rate improvement comes from just removing the
debug.Log calls: every one of those copied a restic.ID to the heap.

greatroar commented:
Pushed a version that has a map, but stores indices in the map instead of pointers to prevent the GC from scanning all its entries.

    name                 old time/op    new time/op    delta
    EncodeIndex/100-8      1.56ms ± 2%    1.48ms ± 3%   -5.37%  (p=0.000 n=10+10)
    EncodeIndex/1000-8     14.5ms ± 2%    13.1ms ± 2%   -9.49%  (p=0.000 n=9+10)
    EncodeIndex/10000-8     120ms ± 2%     116ms ± 2%   -3.58%  (p=0.000 n=10+10)
    
    name                 old alloc/op   new alloc/op   delta
    EncodeIndex/100-8       306kB ± 1%     275kB ± 1%  -10.28%  (p=0.000 n=10+10)
    EncodeIndex/1000-8     3.69MB ±11%    2.88MB ± 5%  -22.07%  (p=0.000 n=10+9)
    EncodeIndex/10000-8    35.9MB ±11%    31.9MB ±10%  -11.13%  (p=0.005 n=10+10)
    
    name                 old allocs/op  new allocs/op  delta
    EncodeIndex/100-8       3.39k ± 0%     2.39k ± 0%  -29.61%  (p=0.000 n=10+10)
    EncodeIndex/1000-8      32.6k ± 0%     22.9k ± 0%  -29.63%  (p=0.000 n=10+9)
    EncodeIndex/10000-8      326k ± 0%      229k ± 0%  -29.71%  (p=0.000 n=10+10)

greatroar commented:
Ugh. If I remove restic.ID.MarshalJSON, I can get even fewer allocations, but then encoding indexes gets slower:

```
name                 old time/op    new time/op    delta
EncodeIndex/100-8      1.48ms ± 3%    1.64ms ± 3%  +10.91%  (p=0.000 n=10+10)
EncodeIndex/1000-8     13.1ms ± 2%    15.9ms ± 8%  +21.08%  (p=0.000 n=10+10)
EncodeIndex/10000-8     116ms ± 2%     147ms ± 3%  +26.83%  (p=0.000 n=10+10)

name                 old alloc/op   new alloc/op   delta
EncodeIndex/100-8       275kB ± 1%     193kB ± 2%  -29.73%  (p=0.000 n=10+10)
EncodeIndex/1000-8     2.88MB ± 5%    2.46MB ±13%  -14.54%  (p=0.000 n=9+10)
EncodeIndex/10000-8    31.9MB ±10%    29.2MB ±25%     ~     (p=0.190 n=10+10)

name                 old allocs/op  new allocs/op  delta
EncodeIndex/100-8       2.39k ± 0%     1.38k ± 0%  -42.13%  (p=0.000 n=10+10)
EncodeIndex/1000-8      22.9k ± 0%     13.3k ± 0%  -42.18%  (p=0.000 n=9+10)
EncodeIndex/10000-8      229k ± 0%      132k ± 0%  -42.28%  (p=0.000 n=10+10)
```

This warrants some more research, but I'll leave that out of this PR.

MichaelEischer reviewed (approved):
LGTM. Thanks for optimizing the method! Although I have to admit that I've been wondering a bit whether the additional complexity (the map now represents an additional indirection) is worth it.
